src/Formula/Types.hs

```diff
@@ -258,26 +258,25 @@ instance Arbitrary Cnf where
     arbitrary = sized cnf
       where
         cnf :: Int -> Gen Cnf
-        cnf 0 = genCnf (0,0) (0,0) []
+        cnf 0 = genCnf (0,0) (0,0) [] True
         cnf n = do
             minLen <- chooseInt (1,n)
             let
               lits = take n ['A'..'Z']
               maxLen = length lits
-            genCnf (1, maxLen ^ (2 :: Int)) (minLen, maxLen) lits
+            genCnf (1, maxLen ^ (2 :: Int)) (minLen, maxLen) lits True
 
 
 
 
 -- | Generates a random cnf satisfying the given bounds
 --   for the amount and the length of the contained clauses.
 --   The used atomic formulas are drawn from the list of chars.
---   Every char from the list will occur at least once in the formula.
-genCnf :: (Int,Int) -> (Int,Int) -> [Char] -> Gen Cnf
-genCnf (minNum,maxNum) (minLen,maxLen) lits = do
+genCnf :: (Int,Int) -> (Int,Int) -> [Char] -> Bool -> Gen Cnf
+genCnf (minNum,maxNum) (minLen,maxLen) lits useAllLiterals = do
```

Comment:
Maybe `useAllLiterals` here would better be named `enforceUsingAllLiterals` or similar?

Because setting it to `False` does not mean that not all literals are used. It only means that it is not enforced that all literals are used.